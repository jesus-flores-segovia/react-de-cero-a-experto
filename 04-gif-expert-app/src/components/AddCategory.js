import { useState } from "react"
import PropTypes from "prop-types";

export const AddCategory = ({setCategories}) => {
    
    // Es importante inicializar el state, para evitar el warning que aparece por consola, que nos avisa de cambios incontrolados en el input
    const [inputValue, setInputValue] = useState();
    
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length !== 0){
            // Recibimos las categories en el callback de la funciÃ³n del useState
            setCategories(categories => [...categories, inputValue]);
            setInputValue("");
        }
        
    };

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};