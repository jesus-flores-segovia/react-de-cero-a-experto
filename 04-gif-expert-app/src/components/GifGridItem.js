import PropTypes from "prop-types";

export const GifGridItem = ({id, title, url}) => {

    return (
        <div className="card animate__animated animate__fadeInLeft">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
};

GifGridItem.propTypes = {
    id: PropTypes.string,
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};