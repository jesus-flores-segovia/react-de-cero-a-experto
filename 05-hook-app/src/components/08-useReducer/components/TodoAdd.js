import React from 'react';
import {useForm} from '../../../hooks/useForm';

export const TodoAdd = ({handleAdd}) => {

    const [{description}, handleInputChanges, reset] = useForm({
        description: ""
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length === 0){
            reset();
            return;
        }

        handleAdd(description);
        reset();
    };

    return (
        <>
        <h4>Add TODO</h4>
        <hr/>
                    
        <form onSubmit={handleSubmit}>
            <input type="text" name="description" className="form-control" placeholder="Learn..." autoComplete="off" onChange={handleInputChanges} value={description}/>
            <div className="d-grid">
                <button type="submit" className="btn btn-outline-primary mt-1">Add</button>
            </div>
        </form>
        </>
    )
}
