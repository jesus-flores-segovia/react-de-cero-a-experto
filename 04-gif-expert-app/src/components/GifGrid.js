import { useState, useEffect } from "react";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({category}) => {
    
    const [images, setImages] = useState([]);

    useEffect(() => {
        getGifs();
    }, []);

    const getGifs = async() => {

        const url = "https://api.giphy.com/v1/gifs/search?api_key=H1fsG7dJKwEzuWy3mD6lXASZNOIU0c0y&q=Rick and Morty&limit=10";
        const response = await fetch(url);
        const {data} = await response.json();

        const gifs = data.map(img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url
            }
        })

        setImages(gifs);
    };

    return (
        <div>
            <h3>{category}</h3>
            {
                images.map(img => (
                    <GifGridItem
                        key={img.id}
                        {...img}
                    />
                ))
            }
        </div>
    )
}
