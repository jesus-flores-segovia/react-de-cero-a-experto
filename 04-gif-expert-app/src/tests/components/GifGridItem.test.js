import { GifGridItem } from "../../components/GifGridItem";
import { shallow } from "enzyme";
import { toHaveClass } from '@testing-library/jest-dom'

describe("Tests inside file GifGridItem.test.js", () => {

    const gifGridItemProps = {
        id: "26BRzozg4TCBXv6QU",
        title: "Flying Harry Potter GIF by The Story Room",
        url: "https://media2.giphy.com/media/26BRzozg4TCBXv6QU/giphy.gif?cid=cc01726bdlkfuc1159oizzekhpham5is3gzwae4myiqeygw2&rid=giphy.gif&ct=g"
    };
    
    let wrapper = shallow(<GifGridItem {...gifGridItemProps}></GifGridItem>);

    beforeEach(() => {
        wrapper = shallow(<GifGridItem {...gifGridItemProps}></GifGridItem>);
    });

    test("'GifGridItem' component must show correctly", () => {

        expect(wrapper).toMatchSnapshot();
    });

    test("'GifGridItem' component must have a title", () => {

        const p = wrapper.find("p");

        expect(p.text()).toBe(gifGridItemProps.title);
    });

    test("'GifGridItem' component must have a html img tag", () => {

        const img = wrapper.find("img");

        expect(img.exists()).toBe(true);
    });

    test("'GifGridItem' component must have a html img tag including src attribute", () => {

        const img = wrapper.find("img");
        const src = img.prop("src");

        expect(src).toBe(gifGridItemProps.url);
    });

    test("'GifGridItem' component must have a html img tag including alt attribute", () => {

        const img = wrapper.find("img");
        const alt = img.prop("alt");

        expect(alt).toBe(gifGridItemProps.title);    
    });

    test("'GifGridItem' component must have a html div tag configured with a CSS animation attribute values as these: 'animate__animated animate__fadeInLeft'", () => {

        const div = wrapper.find("div");

        expect(div.hasClass("animate__animated animate__fadeInLeft")).toBe(true);
    });

});