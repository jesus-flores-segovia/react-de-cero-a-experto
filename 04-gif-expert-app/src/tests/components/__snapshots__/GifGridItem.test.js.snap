// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file GifGridItem.test.js 'GifGridItem' component must show correctly 1`] = `
<div
  className="card animate__animated animate__fadeInLeft"
>
  <img
    alt="Flying Harry Potter GIF by The Story Room"
    src="https://media2.giphy.com/media/26BRzozg4TCBXv6QU/giphy.gif?cid=cc01726bdlkfuc1159oizzekhpham5is3gzwae4myiqeygw2&rid=giphy.gif&ct=g"
  />
  <p>
    Flying Harry Potter GIF by The Story Room
  </p>
</div>
`;
